### PR TITLE
Add modifications to deploy release prompt

### PR DESCRIPTION
**Pull request by alvaro-shopify**
### WHY are these changes introduced?

Follow-up: https://github.com/Shopify/cli/pull/3191

NOTES:

- Rebase follow-up branch before merging it

The previous PR [Adapt deploy release prompts](https://github.com/Shopify/cli/pull/3186) included the following:
- First UX proposal for the `deploy` and `release` commands
- Great refactoring about how to build these prompts

In the meantime the UX proposal has been modified. Furthermore, as the PR was really huge, the implementation needed to add the `config diff` content to the `release` command prompt was delayed to a new PR

### WHAT is this pull request doing?
- Final UX proposal for the `deploy and `release` command. Main differences:
  -  Removed the `Removed` section to display the `configuration` and `regular` deleted app modules. Now they are listed in their own `Configuration` or `Extension` sections
  - Add support for custom styled list items inside the `InfoTable` content (now each item can use its own `color` and `bullet`
  - Add support for a custom text message if the list of items inside the `InfoTable` is empty (to display the message `No changes` without a list styling
- Add the  the `config diff` content to the `release` command prompt. To achieve these the following modifications in the APIs were needed to fetch the `config` content from the `app version` (either the active or the one to be released)
  - [shopify](https://github.com/Shopify/shopify/commit/3a41edb6c1bb4f86b4f842c7fec2eeae6929e100)
  - [partners](https://github.com/Shopify/partners/commit/ccff78c168ee3bb9a764e81387e018dc2b1)
- Fix an issue from the first PR as the `config diff` content displayed in the `deploy` command was using the current remote extensions registrations content instead of the `active version` content.

### How to test your changes?
`api_version` is [versioned](https://github.com/Shopify/shopify/pull/472181) so it doesn't worth adding logic that should be removed really soon. We can wait until that PR is shipped to merge this one
- Create a new `cli` app and run the `p shopify app config link --path /PATH/TO/YOUR/APP` command to link it to a new remote app. 
- Deploy with release `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP` a first version (`A`) with the following content in the `toml` file:
```
name = "<YOUR_NAME>"
client_id = "<YOUR_CLIENT_ID>"
application_url = "https://shopify.dev/apps/default-app-home"
embedded = true

[webhooks]
api_version = "unstable"

[pos]
embedded = false
```
- Deploy with release `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP`a second version (`B`) with the following content in the `toml` file:
```
name = "<YOUR_NAME>"
client_id = "<YOUR_CLIENT_ID>"
application_url = "https://shopify.dev/apps/default-app-home-2"
embedded = false

[webhooks]
api_version = "unstable"

[app_proxy]
url = "https://my-proxy-new.dev"
subpath = "subpath-whatever"
prefix = "apps"
```

#### Scenario 1
- From the partners dashboard `Release` the version `A`
- From the CLI `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP` with release. This should be the output but don't confirm:
```
   ┃  + app_proxy (new)
   ┃  • application_url
   ┃  • embedded
   ┃  - pos (removed)
```

#### Scenario 2
- NOTE: Prompts will always display `┃  + webhooks (new)` and  it's not possible to display the message `No changes`.  The problem is that `api_version` inside `webhooks` is not versioned yet in the server side thus it's not deployed. Each time you run `deploy` command you will see the change. This behaviour will be fixed once 
- From the partners dashboard `Release` the version `B`
- From the CLI `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP`  with release. This should be the output but don't confirm
```
   ┃   + app_proxy (new)
```

#### Scenario 3
- From the CLI `npm run shopify app release --path /PATH/TO/YOUR/APP --version YOUR-VERSION-A` the version `A`.  This should be the output and confirm the release
```
   ┃  + pos (new)
   ┃  • application_url
   ┃  • embedded
   ┃  - app_proxy (removed)
```

#### Scenario 4
- From the CLI `npm run shopify app release --path /PATH/TO/YOUR/APP --version YOUR-VERSION-B` the version `B`.  This should be the output and confirm the release
```
   ┃  + app_proxy (new)
   ┃  • application_url
   ┃  • embedded
   ┃  - pos (removed)
```

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/prompts/deploy-release.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import {buildNURInfoTableSection} from './ui/NURInfoTableSection.js'
```
Comment (alexanderMontague) [inline packages/app/src/cli/prompts/deploy-release.ts:1]: nit, what is `NUR`? "new", "updated", "removed"? Can we de-acronym this for better readability?

File: packages/app/src/cli/prompts/ui/NURInfoTableSection.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+const RemovedItem = {
+  bullet: '-',
+  color: 'red',
+  suffix: '(removed)',
+}
+
+export function buildNURInfoTableSection<T>(section: NURInfoTableSection<T>) {
+  return [
+    ...section.new.map((item) => buildItem(item, NewItem)),
+    ...section.updated.map((item) => item),
```
Comment (alexanderMontague) [inline packages/app/src/cli/prompts/ui/NURInfoTableSection.ts:28]:
```suggestion
    ...section.updated,
```

File: packages/cli-kit/src/private/node/ui/components/List.tsx [hunk 3]
```diff
@@ -25,23 +34,57 @@ const List: FunctionComponent<ListProps> = ({
   color,
   bullet = DOT,
 }): JSX.Element => {
+  function isCustomListItem(item: TokenItem<InlineToken> | CustomListItem): item is CustomListItem {
+    return (item as CustomListItem).item !== undefined
+  }
```
Comment (alexanderMontague) [inline packages/cli-kit/src/private/node/ui/components/List.tsx:39]: why do we have to support `TokenItem<InlineToken>` if we cast and override return type 😅 

edit: actually I see, we can replace this with your new `ListItem` type. I think we can probably do better with the typing here, but I won't hold things up. Let me know if you want to pair on this!

**Review by alexanderMontague — APPROVED** [review]
Just some nits, but the tophat works great! Everything looks good 👍 (minus the always new webhooks config item)

<details>
<summary>tophat output</summary>

image

image

image

image


</details

File: packages/cli-kit/src/private/node/ui/components/List.tsx [hunk 4]
```diff
@@ -25,23 +34,57 @@ const List: FunctionComponent<ListProps> = ({
...
+      color,
+      bullet,
+      ordered,
+      item: item as TokenItem<InlineToken>,
+    }
+    return isCustomListItem(item)
+      ? {
+          ...resolvedItem,
+          ...item,
+        }
```
Comment (isaacroldan) [inline packages/cli-kit/src/private/node/ui/components/List.tsx:53]: `resolvedItem` already includes `item` no? why do you need to merge them here? 🤔

**Pull request merged**